core/corehttp/gateway_handler_unixfs.go

```diff
@@ -30,9 +140,124 @@ func (i *gatewayHandler) serveUnixFs(w http.ResponseWriter, r *http.Request, res
 	// Handling Unixfs directory
 	dir, ok := dr.(files.Directory)
 	if !ok {
-		internalWebError(w, fmt.Errorf("unsupported UnixFs type"))
+		internalWebError(w, fmt.Errorf("unsupported Unixfs type"))
 		return
 	}
 	logger.Debugw("serving unixfs directory", "path", contentPath)
 	i.serveDirectory(w, r, resolvedPath, contentPath, dir, begin, logger)
 }
+
+// redirect returns redirected, newPath (if rewrite), error
+func (i *gatewayHandler) redirect(w http.ResponseWriter, r *http.Request, path ipath.Resolved) (bool, string, error) {
+	node, err := i.api.Unixfs().Get(r.Context(), path)
+	if err != nil {
+		return false, "", fmt.Errorf("could not get redirects file: %v", err)
+	}
+
+	defer node.Close()
+
+	f, ok := node.(files.File)
+
+	if !ok {
+		return false, "", fmt.Errorf("redirect, could not convert node to file")
+	}
+
+	redirs := newRedirs(f)
+
+	// extract "file" part of URL, typically the part after /ipfs/CID/...
+	g := strings.Split(r.URL.Path, "/")
+
+	if len(g) > 3 {
+		filePartPath := "/" + strings.Join(g[3:], "/")
+
+		to, code := redirs.search(filePartPath)
+		if code > 0 {
+			if code == 200 {
+				// rewrite
+				newPath := strings.Join(g[0:3], "/") + "/" + to
+				return false, newPath, nil
+			}
+
+			// redirect
+			http.Redirect(w, r, to, code)
+			return true, "", nil
+		}
+	}
+
+	return false, "", nil
+}
+
+// Returns a resolved path to the _redirects file located in the root CID path of the requested path
+func (i *gatewayHandler) getRedirectsFile(r *http.Request) (ipath.Resolved, error) {
+	// r.URL.Path is the full ipfs path to the requested resource,
+	// regardless of whether path or subdomain resolution is used.
+	rootPath := getRootPath(r.URL.Path)
+	path := ipath.New(gopath.Join(rootPath, "_redirects"))
+	resolvedPath, err := i.api.ResolvePath(r.Context(), path)
+	if err != nil {
+		return nil, err
+	}
+	return resolvedPath, nil
+}
+
+// Returns the root CID path for the given path
+func getRootPath(path string) string {
+	if strings.HasPrefix(path, ipfsPathPrefix) && strings.Count(gopath.Clean(path), "/") >= 2 {
+		parts := strings.Split(path, "/")
+		return gopath.Join(ipfsPathPrefix, parts[2])
+	} else {
+		return ""
+	}
+}
+
+// TODO(JJ): I was thinking about changing this to just look at the root path as well, but the docs say it searches up
```

Comment:
I assume you agree?

Reply:
Okay, makes sense.  So I assume I should keep existing pretty 404 logic, but only fall back to it if there isn't any [netlify style custom 404 page handling](https://docs.netlify.com/routing/redirects/redirect-options/#custom-404-page-handling) entry in the _redirects file.  And only the new `_redirects` file 404 logic will be documented.  👌